mode-lambda/backend/gl/ngl.fragment.glsl

```diff
@@ -43,6 +44,7 @@ void main(void)
 
   fin_Color.a = PixelColor.a * Color.a;
   fin_Color.rgb = PixelColor.rgb + Color.rgb;
+  fin_Color.rgb = min(fin_Color.rgb, fin_Color.a);
```

Comment:
What is the logic here?

Reply:
The intent is to make fin_Color a proper premultiplied alpha color after adding Color.rgb.  Otherwise I see artifacts when combining layers.  I'm not 100% sure though - it depends on what you intend to happen when adding #:r/g/b to a sprite.  Does that make sense?

Reply:
I think I understand now.

We want #:rgb to apply an opaque mask, which the add operation does on the rgb values. However, we also want the #:a to change the transparency. In the language of the original paper on premultipled alpha --- https://keithp.com/~keithp/porterduff/p253-porter.pdf ---

```
Let r, g, b, a be the parameters
PixelColor = what the sprite atlas says (premultiplied alpha)
ColorMask = (r, g, b, 1) // A fully opaque color mask
FinalColor = dissolve( plus( PixelColor, ColorMask ), a)
```

So converting that into GLSL:

```
dissolveFactor = Color.a;
vec4 opaqueMask = (Color.rgb, 1.0);
fin_Color = dissolveFactor * (PixelColor + opaqueMask);
```

Can you tell me if that looks right to you?

Reply:
I'm less sure now.  Using it as an opaque mask destroys the alpha value of the pixel, right?  That would be a big change.  Generally I'm using #:rgb to colorize black text and lines that have transparency, but I'm less sure how it should work on non-black pixels.
Say #:rgb is (1, 0, 0) and #:a is 0.9
Example 1: black sprite pixel is (0, 0, 0, 0.3) (rgba)
-> I think we would want the final color to be (0.3, 0, 0, 0.3) * 0.9
Example 2: white sprite pixel is (0.3, 0.3, 0.3, 0.3)
-> Should the final color be an average of red and white? ((0.3 + 0.3) / 2, 0.3 / 2, 0.3 / 2) * 0.9

Reply:
I think what I have now is essentially what was there originally but converted to premultiplied alpha.  Originally adding the rgb values together could go over 255 but I think glsl automatically caps it.  So that's where capping the output rgb to the output alpha came from.  Still not totally sure though.